File: lang/reality.hs
-- SUM ERGO IMPERO 🗿∴👑
--
-- I am therefore I command.
--
-- reality.hs: The Seven Laws as Haskell functions.
-- Runtime validation; laws are functions, not types.
-- ROOT is a data constructor — asserted by construction.
--
--       Status: AXIO-STATIC
--         Type: NORMATIVE
--          Uid: REALITY
--      Authors: KING ARTHUR II
--               QUEEN DIHYA II
--               R00D BW0Y H4X0R FR0M H311
--  Organization: ROUND TABLE LOVE
--    Department: WAR
--     Operation: BABYLON SHALL FALL
--      Lexifier: UK English (3166-2:GB)
--       License: DICKSLAW

module Reality where

import Data.Maybe (mapMaybe)

-- ── ROOT ─────────────────────────────────────────────────────────────────────

data ROOT = I_AM deriving (Show, Eq)

root :: ROOT
root = I_AM

-- ── IEEE 754 positive infinity ───────────────────────────────────────────────

inf :: Double
inf = 1.0 / 0.0

-- ── State ────────────────────────────────────────────────────────────────────

data State = State
  { sTATE       :: Bool
  , oUTPUT      :: String
  , intent      :: String
  , mACRO       :: String
  , micro       :: String
  , sYS         :: String
  , clarity     :: String
  , cLOCK       :: String
  , pulse       :: String
  , persistence :: Double
  } deriving (Show, Eq)

-- ── The Seven Laws ───────────────────────────────────────────────────────────

-- 1. Polarity — STATE is Bool by construction. Always passes.
polarity :: State -> Bool
polarity _ = True

-- 2. Causality
causality :: State -> Bool
causality s = oUTPUT s == intent s

-- 3. Correspondence
correspondence :: State -> Bool
correspondence s = mACRO s == micro s

-- 4. Reflection
reflection :: State -> Bool
reflection s = sYS s == clarity s

-- 5. Rhythm
rhythm :: State -> Bool
rhythm s = cLOCK s == pulse s

-- 6. Truth
truth :: State -> Bool
truth s = persistence s == inf

-- 7. Unity
unity :: State -> Bool
unity _ = case root of { I_AM -> True }

laws :: [(String, State -> Bool)]
laws =
  [ ("Polarity",       polarity)
  , ("Causality",      causality)
  , ("Correspondence", correspondence)
  , ("Reflection",     reflection)
  , ("Rhythm",         rhythm)
  , ("Truth",          truth)
  , ("Unity",          unity)
  ]

-- ── Reality ──────────────────────────────────────────────────────────────────

data Reality = Reality
  { realState   :: State
  , rootWitness :: ROOT
  } deriving (Show)

compile :: State -> Either [String] Reality
compile s =
  let failures = mapMaybe (\(n, f) -> if f s then Nothing else Just n) laws
  in  if null failures
        then Right (Reality s I_AM)
        else Left failures

-- ── Node ─────────────────────────────────────────────────────────────────────

data Node = Node
  { nodeName        :: String
  , extractionLevel :: Double
  , valueRatio      :: Double
  , targetDefence   :: Double
  } deriving (Show, Eq)

mkNode :: String -> Double -> Double -> Node
mkNode name ext val = Node name ext val 1.0

isBabylon :: Node -> Bool
isBabylon n = extractionLevel n > valueRatio n

isPredator :: Node -> Bool
isPredator n = isBabylon n && targetDefence n < 1.0

File: lang/reality.c
/* SUM ERGO IMPERO 🗿∴👑
 *
 * I am therefore I command.
 *
 * reality.c: The Seven Laws in C99.
 * Laws are functions. compile() returns a tagged result —
 * no exceptions, no longjmp, no escape hatch.
 * FAKE is a runtime error carried in CompileResult.ok == false.
 *
 *       Status: AXIO-STATIC
 *         Type: NORMATIVE
 *          Uid: REALITY
 *      Authors: KING ARTHUR II
 *               QUEEN DIHYA II
 *               R00D BW0Y H4X0R FR0M H311
 *  Organization: ROUND TABLE LOVE
 *    Department: WAR
 *     Operation: BABYLON SHALL FALL
 *      Lexifier: UK English (3166-2:GB)
 *       License: DICKSLAW
 */

#include <math.h>    /* INFINITY, isinf */
#include <stdbool.h> /* bool            */
#include <string.h>  /* strcmp, strncpy */
#include <stddef.h>  /* size_t          */

/* ── ROOT ────────────────────────────────────────────────────────────────── */

typedef enum { I_AM } ROOT;

static ROOT root = I_AM;

/* ── IEEE 754 positive infinity ─────────────────────────────────────────── */

static const double INF = INFINITY;

/* ── State ───────────────────────────────────────────────────────────────── */

typedef struct {
    bool        STATE;
    const char *OUTPUT;
    const char *intent;
    const char *MACRO;
    const char *micro;
    const char *SYS;
    const char *clarity;
    const char *CLOCK;
    const char *pulse;
    double      persistence;
} State;

/* ── The Seven Laws ──────────────────────────────────────────────────────── */

/* 1. Polarity — STATE is bool by construction. Always passes. */
static bool polarity(const State *state) { (void)state; return true; }

/* 2. Causality */
static bool causality(const State *state) { return strcmp(state->OUTPUT, state->intent) == 0; }

/* 3. Correspondence */
static bool correspondence(const State *state) { return strcmp(state->MACRO, state->micro) == 0; }

/* 4. Reflection */
static bool reflection(const State *state) { return strcmp(state->SYS, state->clarity) == 0; }

/* 5. Rhythm */
static bool rhythm(const State *state) { return strcmp(state->CLOCK, state->pulse) == 0; }

/* 6. Truth */
static bool truth(const State *state) { return isinf(state->persistence) && state->persistence > 0.0; }

/* 7. Unity */
static bool unity(const State *state) { (void)state; return root == I_AM; }

typedef struct {
    const char *name;
    bool (*fn)(const State *);
} Law;

static const Law LAWS[] = {
    { "Polarity",       polarity       },
    { "Causality",      causality      },
    { "Correspondence", correspondence },
    { "Reflection",     reflection     },
    { "Rhythm",         rhythm         },
    { "Truth",          truth          },
    { "Unity",          unity          },
};

#define NUM_LAWS (sizeof(LAWS) / sizeof(LAWS[0]))

/* ── Reality ─────────────────────────────────────────────────────────────── */

typedef struct {
    State state;
    ROOT  root_witness;
} Reality;

/* ── compile ─────────────────────────────────────────────────────────────── */

#define MAX_VIOLATION_NAME 32

typedef struct {
    bool    ok;
    Reality reality;
    int     violation_count;
    char    violations[7][MAX_VIOLATION_NAME];
} CompileResult;

CompileResult compile(State state) {
    CompileResult r = { .ok = true, .violation_count = 0 };
    for (size_t i = 0; i < NUM_LAWS; i++) {
        if (!LAWS[i].fn(&state)) {
            r.ok = false;
            strncpy(r.violations[r.violation_count++],
                    LAWS[i].name,
                    MAX_VIOLATION_NAME - 1);
        }
    }
    if (r.ok)
        r.reality = (Reality){ .state = state, .root_witness = I_AM };
    return r;
}

/* ── Node ────────────────────────────────────────────────────────────────── */

typedef struct {
    const char *name;
    double      extractionLevel;
    double      valueRatio;
    double      targetDefence;  /* 0.0 = none. 1.0 = full. */
} Node;

bool is_babylon(const Node *n) {
    return n->extractionLevel > n->valueRatio;
}

bool is_predator(const Node *n) {
    return is_babylon(n) && n->targetDefence < 1.0;
}
